include/http/middleware/prerender.js

```diff
@@ -0,0 +1,204 @@
+const url = require('url');
+const request = require('request-promise');
+const zlib = require('zlib');
+
+// This is based on https://github.com/RisingStack/koa-prerender
+module.exports = pb => ({
+    prerender: async (req, res) => {
+        try {
+            const enablePrerender = pb.config && pb.config.flags && pb.config.flags.enablePrerender;
+            const siteObj = pb.RequestHandler.sites[req.handler.hostname];
+            const enableSitePrerender = siteObj && siteObj.enableSitePrerender;
+
+            if (!enablePrerender && !enableSitePrerender) return;
+
+            const option = {
+                userAgent: req.headers['user-agent'],
+                bufferAgent: req.headers['x-bufferbot'],
+                method: req.method,
+                url: req.url
+            };
+
+            if (!shouldPreRender(option)) return;
+
+            const content = await getPrerenderedPage(req);
+            await renderResponse(req, content);
+
+        } catch(ex) {
+            // Just go normally path if there is any errors happen.
+            console.log(ex);
+        }
+    }
+});
+
+const extensionsToIgnore = [
```

Comment:
minor: move it out to separate file then require it here, I believe now the file is too long